core/src/cluster_info_repair_listener.rs

```diff
@@ -0,0 +1,683 @@
+                let last_update_ts = Self::get_last_ts(peer.id, peer_roots);
+                let my_root =
+                    Self::update_my_gossiped_root(&my_id, cluster_info, &mut my_gossiped_root);
+                {
+                    let r_cluster_info = cluster_info.read().unwrap();
+
+                    // Update our local map with the updated peers' information
+                    if let Some((peer_epoch_slots, ts)) =
+                        r_cluster_info.get_epoch_state_for_node(&peer.id, last_update_ts)
+                    {
+                        // Following logic needs to be fast because it holds the lock
+                        // preventing updates on gossip
+                        peer_roots.insert(peer.id, (ts, peer_epoch_slots.root));
+                        if Self::should_repair_peer(
+                            my_root,
+                            peer_epoch_slots.root,
+                            &epoch_schedule,
+                            NUM_BUFFER_SLOTS,
+                        ) {
+                            // Clone out EpochSlots structure to avoid holding lock on gossip
+                            peers_needing_repairs.insert(peer.id, peer_epoch_slots.clone());
+                        }
+                    }
+                }
+            }
+
+            // After updating all the peers, send out repairs to those that need it
+            let _ = Self::serve_repairs(
+                &my_id,
+                blocktree,
+                peer_roots,
+                &peers_needing_repairs,
+                &socket,
+                cluster_info,
+                &epoch_schedule,
+                &mut my_gossiped_root,
+            );
+
+            sleep(Duration::from_millis(REPAIRMEN_SLEEP_MILLIS as u64));
+        }
+    }
+
+    fn serve_repairs(
+        my_id: &Pubkey,
+        blocktree: &Blocktree,
+        peer_roots: &HashMap<Pubkey, (u64, u64)>,
+        repairees: &HashMap<Pubkey, EpochSlots>,
+        socket: &UdpSocket,
+        cluster_info: &Arc<RwLock<ClusterInfo>>,
+        epoch_schedule: &EpochSchedule,
+        my_gossiped_root: &mut u64,
+    ) -> Result<()> {
+        for (repairee_id, repairee_epoch_slots) in repairees {
+            let repairee_root = repairee_epoch_slots.root;
+
+            let repairee_tvu = {
+                let r_cluster_info = cluster_info.read().unwrap();
+                let contact_info = r_cluster_info.get_contact_info_for_node(repairee_id);
+                contact_info.map(|c| c.tvu)
+            };
+
+            if let Some(repairee_tvu) = repairee_tvu {
+                // For every repairee, get the set of repairmen who are responsible for
+                let mut eligible_repairmen = Self::find_eligible_repairmen(
+                    repairee_root,
+                    peer_roots,
+                    epoch_schedule,
+                    NUM_BUFFER_SLOTS,
+                );
+
+                eligible_repairmen.push(my_id);
```

Comment:
why wouldn't the find_eligible_repairmen() function not include my_id?  why push my_id unconditionally?  if my_id is not an eligible repairman, we're done here, right?